assets/helpers/paymentIntegrations/paymentApi.js

```diff
@@ -0,0 +1,202 @@
+// @flow
+
+import { routes } from 'helpers/routes';
+import { addQueryParamsToURL } from 'helpers/url';
+import {
+  type ReferrerAcquisitionData,
+  type PaymentAPIAcquisitionData,
+  type OphanIds,
+  type AcquisitionABTest,
+} from 'helpers/tracking/acquisitions';
+import { type Csrf as CsrfState } from 'helpers/csrf/csrfReducer';
+import { type BillingPeriod, type Contrib } from 'helpers/contributions';
+import { type IsoCurrency } from 'helpers/internationalisation/currency';
+import { type UsState, type CaState, type IsoCountry } from 'helpers/internationalisation/country';
+import { pollP, logP } from 'helpers/promise';
+
+import * as cookie from 'helpers/cookie';
+
+// ----- Types ----- //
+
+type OneOffFields = {|
+  paymentData: {
+    currency: IsoCurrency,
+    amount: number,
+    token: string,
+    email: string
+  },
+  acquisitionData: PaymentAPIAcquisitionData,
+|};
+
+type RegularContribution = {|
+  amount: number,
+  currency: string,
+  billingPeriod: BillingPeriod,
+|};
+
+type PayPalDetails = {| baid: string |};
+
+type StripeDetails = {| stripeToken: string |};
+
+type PaymentDetails = PayPalDetails | StripeDetails;
+
+type RegularFields = {|
+  firstName: string,
+  lastName: string,
+  country: IsoCountry,
+  state: UsState | CaState | null,
+  email: string,
+  contribution: RegularContribution,
+  paymentFields: PaymentDetails,
+  ophanIds: OphanIds,
+  referrerAcquisitionData: ReferrerAcquisitionData,
+  supportAbTests: AcquisitionABTest[],
+|};
+
+type PaymentFields 
+  = {| tag: 'oneoff', fields: OneOffFields |}
+  | {| tag: 'regular', fields: RegularFields |}
+  ;
+
+type Credentials = 'omit' | 'same-origin' | 'include';
+
+export type Token
+  = {| tag: 'Stripe', token: string |}
+  | {| tag: 'PayPal', token: string |}
+  | {| tag: 'DirectDebit', accountHolderName: string, sortCode: string, accountNumber: string |}
+  ;
+
+export type PaymentResult
+  = {| tag: 'success' |}
```

Comment:
I like this idea 👍

However, perhaps it would be clearer if we went for more meaningful names (e.g. _paymentStatus_ instead of _tag_)?